### PR TITLE
Seccomp: unify high level filter building for basic and advanced mode

### PR DESCRIPTION
**Pull request by aghecenco**
Issue #, if available: #920, #972

Description of changes:

* unified the high-level declaration of seccomp filters for `basic` mode and `advanced` mode by leveraging the `SeccompFilterContext` structure, which translates itself into BPF statements, instead of manually composing them for `basic` mode;
* changed the way the `basic` filter list is built - eliminated the static array of allowed syscalls and switched to using the default filter context, which is modified in place so as to allow any argument values for the syscalls it contains;
* eliminated the `SeccompLevel` enum and simplified the code a bit;
* inlined BPF statement builder functions;
* added macros and inline helper functions to improve the code readabilty in building the filters;
* whitelisted `SYS_getrandom` because the MMDS integration test seems to call it (caused the PR to fail because of seccomp violation);
* added some more unit tests for the seccomp code (and not only).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by aghecenco** [comment]
This PR keeps failing the `test_vsock_ping_pong` test. It's timing out while trying to initiate the ssh connection with the guest. It doesn't look like it was done building the auxiliary binaries before starting to attempt connecting; the `test_vsock` binary was missing, and `top` showed 2 `rustc` processes taking up the most cpu. I think the fixture that builds the aux binaries (now also in charge of building the demo seccomp binaries) is taking too long to build (for reasons yet undiscovered) and affecting this unrelated test.

As a workaround, I can revert from using the `aux_bin_paths` fixture for the seccomp binaries to using a dedicated one, to prevent collateral damage.

**Comment by aghecenco** [comment]
After a closer look, the merge between 9af9fc8e3dd3aec9fb6f13fd5c4eb58c9d1a3c98 and this PR went wrong and the ioctl filters were wrong.

File: vmm/src/default_syscalls/x86_64.rs [hunk 1]
```diff
@@ -324,10 +263,7 @@ pub fn default_context() -> Result<SeccompFilterContext, Error> {
                     ],
                 ),
             ),
-            (
-                libc::SYS_munmap,
-                (0, vec![SeccompRule::new(vec![], SeccompAction::Allow)]),
-            ),
+            allow_syscall(libc::SYS_munmap),
             (
                 libc::SYS_open,
```
Comment (blitz) [inline vmm/src/default_syscalls/x86_64.rs]: The filter for the open system call seems broken. As far as I understand it, any open call will match the first rule and all the other flags checking doesn't matter.

This is not an issue in your PR, but since you are here, you might want to fix or at least remove the unused code.

Comment (blitz) [reply vmm/src/default_syscalls/x86_64.rs]: I've opened #972.

Comment (aghecenco) [reply vmm/src/default_syscalls/x86_64.rs]: Nice catch, thanks! The rules following `vec![]` are indeed pointless.

**Comment by aghecenco** [comment]
Discussed offline with @dhrgit and the next actions here would be to:
* drop the `priority` as we don't actually use it; if we want to, we can take care to sort the rules when the filter is built so that the high priority ones end up earlier in the filter.
* add a shorthand for rule building similar to `allow_syscall` but for syscalls with arguments too. The code looks weird now with a mix of `allow_syscalls` and manually constructed tuples.
* rethink the `context.allow_all()` construction. What we want is to have a single list of syscalls, and 2 generators that use it to build filters, without overcomplicating the code. I don't have a solution for this yet.

@dhrgit I'm sure there was more but I forgot :smile: Please add comments for anything else you saw that doesn't quite fit.

File: seccomp/src/lib.rs [hunk 2]
```diff
@@ -785,12 +774,12 @@ impl SeccompFilterContext {
     /// * `default_action` - Action taken for all syscalls that do not match any rule.
     ///
     pub fn new(
-        rules: HashMap<i64, (i64, Vec<SeccompRule>)>,
+        rules: BTreeMap<i64, Vec<SeccompRule>>,
```
Comment (dhrgit) [inline seccomp/src/lib.rs]: As I understand, this will order the rules by the syscall no (in the BPF code). Why not use a plain vector for sending/storing the ruleset list? Then, the order of the elements in the vector would be reflected in the BPF code, and the caller would have control over which ruleset gets checked first.

I.e., if we used a struct to represent the ruleset:

```rust
struct SeccompRuleset {
	syscall_no: i64,
	rules: Vec<SeccompRule>
}
```

and a vector to store all the rulesets, then, we could define the complex set of rules once:

```rust
advanced_ruleset_list = vec![
	allow_syscall_if(most_frequent_syscall, AND_OR_MACROS_AND_STUFF),
	allow_syscall_if(second_most_frequent_syscall, ANOTHER_SET_OF_AND_OR_MACROS),
	...
]
```

and choose how to apply them, based on the seccomp level the user requested:

```rust
ruleset_list = match seccomp_level {
	SECCOMP_LEVEL_ADVANCED => advanced_ruleset_list,
	SECCOMP_LEVEL_BASIC => advanced_ruleset_list.map(|rs| allow_syscall(rs.syscall_no))
}

SeccompFilterContext::new(ruleset_list)
```

Also, this would not require `SeccompFilterContext::allow_all()`, which I find a bit confusing - setting some rules, then resetting them all.

File: seccomp/src/lib.rs [hunk 3]
```diff
@@ -942,27 +922,11 @@ impl SeccompFilterContext {
 ///
 /// * `level` - Filtering level.
 ///
-pub fn setup_seccomp(level: SeccompLevel) -> Result<()> {
+pub fn setup_seccomp(context: SeccompFilterContext) -> Result<()> {
```
Comment (dhrgit) [inline seccomp/src/lib.rs]: Why not have this as part of the filter struct? I.e. `SeccompFilterContext::apply()`.

File: seccomp/src/lib.rs [hunk 4]
```diff
@@ -407,10 +376,31 @@ pub struct SeccompRule {
...
+#[inline(always)]
+pub fn allow_syscall_if(syscall_number: i64, rules: Vec<SeccompRule>) -> SyscallRuleSet {
+    (syscall_number, rules)
+}
+
 /// Filter context containing rules assigned to syscall numbers.
 pub struct SeccompFilterContext {
-    /// Hash map, mapping a priority and a chain of rules to a syscall number.
-    rules: HashMap<i64, (i64, Vec<SeccompRule>)>,
+    /// Hash map, mapping a chain of rules to a syscall number.
```
Comment (dhrgit) [inline seccomp/src/lib.rs]: I know this has been called this since forever, but, since you're changing stuff here, can we call this `SeccompFilter`? I don't get the `Context` part - sounds like a catch-all name, like saying `SeccompFilterStuff`.

File: seccomp/src/lib.rs (hunk 2 above)
Comment (aghecenco) [reply seccomp/src/lib.rs]: Yup now I remember we talked about these issues, then my memory failed me :smiley:

Comment (aghecenco) [reply seccomp/src/lib.rs]: I spent the better part of today trying out this approach and I'd like to keep the `map`. The `vec` doesn't give us a good way to get to the rules for a syscall, and it's tedious (and overcomplicates the code) to add functionality for modifying in place, **or** inserting if no elements are modifiable (see the `add_rules` function). Moreover, we weren't using the priority approach even when we had support for it; I see no reason to add yet another layer of indirection so that we can have a functionality we've never used.

If you feel strongly about this, I will add another `struct` that wraps a `vec`, so that we can keep the mutation benefits of a `map` while also preserving order (although I'd rather not).

File: tests/integration_tests/security/test_seccomp.py [hunk 5]
```diff
@@ -11,11 +11,11 @@
 import host_tools.cargo_build as host  # pylint:disable=import-error
 
 
-@pytest.fixture
-def tmp_basic_jailer(test_session_root_path):
-    """Build `demo_basic_jailer`, required for the basic seccomp tests.
+@pytest.fixture(scope='session')
```
Comment (dianpopa) [inline tests/integration_tests/security/test_seccomp.py]: nit: rephrase the the commits that change the seccomp integration tests in a similar way. i.e:
* seccomp integ test: ...
* seccomp integ tests: ...
* seccomp tests: ...

File: tests/integration_tests/security/demo_advanced_seccomp/src/bin/demo_advanced_jailer.rs [hunk 6]
```diff
@@ -4,165 +4,51 @@
 extern crate libc;
 extern crate seccomp;
 
+mod seccomp_rules;
```
Comment (dianpopa) [inline tests/integration_tests/security/demo_advanced_seccomp/src/bin/demo_advanced_jailer.rs]: nit: full stop at end of commit message.

File: tests/conftest.py [hunk 7]
```diff
@@ -228,9 +234,54 @@ def aux_bin_paths(test_session_root_path):
...
     )
+
+    seccomp_build_path = os.path.join(
+        test_session_root_path,
+        build_tools.CARGO_RELEASE_REL_PATH
+    )
+
+    build_tools.cargo_build(seccomp_build_path,
+                            extra_args='--release',
+                            src_dir='integration_tests/security/demo_seccomp')
```
Comment (dianpopa) [inline tests/conftest.py]: Let us use a constant for this path.

File: tests/conftest.py [hunk 8]
```diff
@@ -213,8 +213,14 @@ def _gcc_compile(src_file, output_file):
 def aux_bin_paths(test_session_root_path):
     """Build external tools.
 
-    They currently consist of a binary that can properly use the clone()
-    syscall, and a very simple vsock client/server application.
+    They currently consist of:
```
Comment (dianpopa) [inline tests/conftest.py]: nit: The commit message starts as an enum with only one item. I would leave it as a simple sentence.

File: vmm/src/default_syscalls/aarch64.rs [hunk 9]
```diff
@@ -1,22 +1,16 @@
...
-        seccomp::SeccompAction::Trap,
-    )
-    .unwrap())
+pub fn default_filter() -> Result<SeccompFilter, Error> {
+    Ok(SeccompFilter::new(vec![].into_iter().collect(), SeccompAction::Trap).unwrap())
 }
 
 /// Applies the configured level of seccomp filtering to the current thread.
+#[allow(unused_variables)]
 pub fn set_seccomp_level(seccomp_level: u32) -> Result<(), Error> {
```
Comment (dianpopa) [inline vmm/src/default_syscalls/aarch64.rs]: You can also do it like this:
```suggestion
pub fn set_seccomp_level(_seccomp_level: u32) -> Result<(), Error> {
```

File: vmm/src/default_syscalls/x86_64.rs [hunk 10]
```diff
@@ -279,55 +279,7 @@ pub fn default_context() -> Result<SeccompFilterContext, Error> {
...
-                                1,
-                                SeccompCmpOp::Eq,
-                                O_RDONLY | O_NONBLOCK | O_CLOEXEC,
-                            )?],
-                            SeccompAction::Allow,
-                        ),
-                    ],
-                ),
-            ),
+            allow_syscall(libc::SYS_open),
```
Comment (dianpopa) [inline vmm/src/default_syscalls/x86_64.rs]: Are we sure we want to relax the allowed parameters for the SYS_open syscall? 
I am trying to understand why keeping the `SeccompRules` for tightening the allowed parameters and remove the one that makes them useless (i.e `SeccompRule::new(vec![], SeccompAction::Allow)`) was not the chosen approach?

File: seccomp/src/lib.rs [hunk 11]
```diff
@@ -142,7 +142,7 @@
 //!     }
 //! }
 //!
-//! fn gen_rules() -> Vec<SyscallRuleSetWithPriority> {
+//! fn gen_rules() -> Vec<SyscallRuleSet> {
```
Comment (dianpopa) [inline seccomp/src/lib.rs]: * In the first commit of this PR the `Vec<(i64, Vec<SeccompRule>)>` is replaced  with newly added type `SyscallRuleSetWithPriority`. In this commit the `SyscallRuleSetWithPriority` is replaced with `SyscallRuleSet` and priority is removed. It is rather confusing this way. What is the reason behind this? Did you want to do it incrementally to make it easier to understand the changes or is it a miss?
* Upon reading the commit message I was under the impression that there was a connection between allow_syscall_if and the removal of the priority. However, after reviewing there seems to be none, right?

File: seccomp/src/lib.rs [hunk 12]
```diff
@@ -378,24 +376,31 @@ pub struct SeccompRule {
...
+#[inline(always)]
+pub fn allow_syscall_if(syscall_number: i64, rules: Vec<SeccompRule>) -> SyscallRuleSet {
+    (syscall_number, rules)
+}
+
 /// Filter context containing rules assigned to syscall numbers.
 pub struct SeccompFilterContext {
-    /// Hash map, mapping a priority and a chain of rules to a syscall number.
-    rules: HashMap<i64, (i64, Vec<SeccompRule>)>,
+    /// Hash map, mapping a chain of rules to a syscall number.
```
Comment (dianpopa) [inline seccomp/src/lib.rs]: Comment still refers to `rules`  as being a "Hash map".

File: seccomp/src/lib.rs [hunk 13]
```diff
@@ -378,24 +376,31 @@ pub struct SeccompRule {
...
         syscall_number,
-        (0, vec![SeccompRule::new(vec![], SeccompAction::Allow)]),
+        vec![SeccompRule::new(vec![], SeccompAction::Allow)],
     )
 }
 
+/// Builds the (syscall, rules) tuple for allowing a syscall with certain arguments.
+///
+#[inline(always)]
+pub fn allow_syscall_if(syscall_number: i64, rules: Vec<SeccompRule>) -> SyscallRuleSet {
```
Comment (dianpopa) [inline seccomp/src/lib.rs]: Am I mistaken or `allow_syscall_if` is only used in tests in this commit?

File: vmm/src/default_syscalls/x86_64.rs [hunk 14]
```diff
@@ -239,10 +179,7 @@ pub fn default_context() -> Result<SeccompFilterContext, Error> {
                 ),
             ),
             (libc::SYS_ioctl, (0, create_ioctl_seccomp_rule()?)),
-            (
-                libc::SYS_lseek,
-                (0, vec![SeccompRule::new(vec![], SeccompAction::Allow)]),
-            ),
+            allow_syscall(libc::SYS_lseek),
             #[cfg(musl)]
```
[the remark this answers is not in the document]
Comment (aghecenco) [reply vmm/src/default_syscalls/x86_64.rs]: aka "the reason behind release 0.15.2" :laughing: Will be fixed by rebase.

File: vmm/src/default_syscalls/x86_64.rs (hunk 10 above)
Comment (aghecenco) [reply vmm/src/default_syscalls/x86_64.rs]: Because `open` gets called with so many combinations of flags `( O_RDONLY or O_RDWR ) [ | O_CLOEXEC ] [ | O_NONBLOCK ]` (possibly more that we haven't discovered yet) that there doesn't seem to be much to gain from overcomplicating the filter.

File: tests/integration_tests/security/demo_advanced_seccomp/src/bin/demo_advanced_jailer.rs [hunk 15]
```diff
@@ -4,165 +4,51 @@
 extern crate libc;
```
[the remark this answers is not in the document]
Comment (aghecenco) [reply tests/integration_tests/security/demo_advanced_seccomp/src/bin/demo_advanced_jailer.rs]: Cargo builds binaries from whatever it finds in `/bin`. Without `/bin`, it would try to pack it as a crate.

File: seccomp/src/lib.rs (hunk 11 above)
Comment (aghecenco) [reply seccomp/src/lib.rs]: 1. I removed the priority after a brief review session with @dhrgit during which we came to the conclusion that we weren't using priorities at all, so this commit went on top of the initial "solution". I tried to squash the 2 commits you identified that alter the same data structure, but there were too many unrelated changes to sweep through and I dropped the effort.
1. There is no connection. During the same review session we agreed to add `allow_syscall_if` to improve readability as there was a mix of `allow_syscall` and explicitly built tuples.

File: seccomp/src/lib.rs (hunk 2 above)
Comment (dhrgit) [reply seccomp/src/lib.rs]: I see what you mean, but I found that `add_rules` was only used within tests. Also, I don't think there's anything preventing us from having multiple rulesets for the same syscall.

Anyway, here's the idea I was talking about, in code form: [seccomp-vec](https://github.com/dhrgit/firecracker/blob/seccomp-vec/seccomp/src/lib.rs)

I think there's also some confusion caused by how we named things in that crate (e.g. rules, conditions, etc), since they don't accurately depict the logical/functional flow. I think there is still room for simplification, but this PR may already become too complex itself. Anyway, let me know what you think.

Comment (aghecenco) [reply seccomp/src/lib.rs]: Please open issues for any further simplifications you have in mind. This PR set out to fix one problem, and has already become hard to follow and to review because of refactoring iterations. At this point, more changes would only make life difficult for everyone.

File: tests/integration_tests/security/test_seccomp.py [hunk 16]
```diff
@@ -20,7 +20,7 @@ def tmp_basic_jailer(test_session_root_path):
     binaries_srcdir = os.path.normpath(
         os.path.join(
             os.getcwd(),
-            'integration_tests/security/demo_advanced_seccomp/'
+            'integration_tests/security/demo_seccomp/'
```
[the remark this answers is not in the document]
Comment (aghecenco) [reply tests/integration_tests/security/test_seccomp.py]: This path is removed in a later commit, which introduces the `aux_bin_paths` fixture in the seccomp test.

File: seccomp/src/lib.rs [hunk 17]
```diff
@@ -942,27 +933,11 @@ impl SeccompFilterContext {
...
-            for &syscall in allowed_syscalls {
-                filters.extend(ALLOW_SYSCALL(syscall));
-            }
-            filters.extend(SIGNAL_PROCESS());
-        }
-        SeccompLevel::None => {
-            return Ok(());
-        }
-    }
+    filters.extend(context.into_bpf().map_err(|_| Error::Load(libc::EINVAL))?);
```
[the remark this answers is not in the document]
Comment (aghecenco) [reply seccomp/src/lib.rs]: Added an `IntoBpf` enum variant, and bonus `Display` trait.

File: seccomp/src/lib.rs [hunk 18]
```diff
@@ -1034,29 +1012,12 @@ fn VALIDATE_ARCHITECTURE() -> Vec<sock_filter> {
...
-    vec![
-        BPF_JUMP(BPF_JMP + BPF_JEQ + BPF_K, syscall_number as u32, 0, 1),
-        BPF_STMT(BPF_RET + BPF_K, SECCOMP_RET_ALLOW),
-    ]
-}
-
-/// Builds a sequence of BPF instructions that emit `SIGSYS` when a syscall is denied.
-///
-#[allow(non_snake_case)]
-fn SIGNAL_PROCESS() -> Vec<sock_filter> {
```
[the remark this answers is not in the document]
Comment (aghecenco) [reply seccomp/src/lib.rs]: Rebased so that every commit passes the unit tests.

File: seccomp/src/lib.rs (hunk 2 above)
Comment (dhrgit) [reply seccomp/src/lib.rs]: Okay then - let's merge this as it is now. I'll open an issue for whenever we find time to take this on.
